File: restdoc-rpc-client/restdoc-restweb-client/src/main/kotlin/restdoc/client/restweb/RestWebInvokerImpl.java
package restdoc.client.restweb;

import org.jetbrains.annotations.NotNull;
import org.springframework.core.env.Environment;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.HttpServerErrorException;
import org.springframework.web.client.RestClientResponseException;
import org.springframework.web.client.RestTemplate;
import restdoc.client.api.Invoker;
import restdoc.client.api.model.InvocationResult;
import restdoc.client.api.model.RestWebInvocation;
import restdoc.client.api.model.RestWebInvocationResult;

import java.util.AbstractMap;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * RestWebInvokerImpl
 */
public class RestWebInvokerImpl implements Invoker<RestWebInvocation> {

    private final RestTemplate restTemplate;

    private final int port;

    private final String contextPath;


    public RestWebInvokerImpl(Environment environment, RestTemplate restTemplate) {
        this.restTemplate = restTemplate;
        this.contextPath = environment.getProperty("server.servlet.context-path", "");
        this.port = Integer.parseInt(environment.getProperty("server.port", "8080"));
    }

    @NotNull
    @Override
    public InvocationResult rpcInvoke(@NotNull RestWebInvocation invocation) {
        String url = autocompleteURL(invocation.url);
        HttpHeaders requestHeaders = new HttpHeaders();
        invocation.requestHeaders.forEach(requestHeaders::addAll);
        HttpEntity httpEntity = new HttpEntity(invocation.requestBody, requestHeaders);

        ResponseEntity<Object> responseEntity = null;

        try {
            responseEntity = restTemplate.exchange(url, HttpMethod.valueOf(invocation.method), httpEntity, Object.class, invocation.uriVariable);

            Map<String, List<String>> responseHeaders = responseEntity.getHeaders().entrySet()
                    .stream()
                    .map(hd -> new AbstractMap.SimpleEntry<>(hd.getKey(), hd.getValue()))
                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));

            return new RestWebInvocationResult(
                    true,
                    null,
                    invocation,
                    responseEntity.getStatusCodeValue(),
                    responseHeaders,
                    responseEntity.getBody());

        } catch (RestClientResponseException e) {
            String exceptionMessage;
            if (e instanceof HttpServerErrorException.BadGateway)
                exceptionMessage = "BadGateway";
            else if (e instanceof HttpClientErrorException.BadRequest)
                exceptionMessage = "BadRequest";
            else if (e instanceof HttpClientErrorException.Conflict)
                exceptionMessage = "Conflict";
            else if (e instanceof HttpClientErrorException.Forbidden)
                exceptionMessage = "Forbidden";
            else if (e instanceof HttpServerErrorException.GatewayTimeout)
                exceptionMessage = "GatewayTimeout";
            else if (e instanceof HttpClientErrorException.Gone)
                exceptionMessage = "Gone";
            else if (e instanceof HttpClientErrorException.NotFound)
                exceptionMessage = "NotFound";
            else if (e instanceof HttpClientErrorException.MethodNotAllowed)
                exceptionMessage = "MethodNotAllowed";
            else if (e instanceof HttpClientErrorException.NotAcceptable)
                exceptionMessage = "NotAcceptable";
            else if (e instanceof HttpClientErrorException.UnsupportedMediaType)
                exceptionMessage = "UnsupportedMediaType";
            else if (e instanceof HttpClientErrorException.UnprocessableEntity)
                exceptionMessage = "UnprocessableEntity";
            else if (e instanceof HttpClientErrorException.TooManyRequests)
                exceptionMessage = "TooManyRequests";
            else if (e instanceof HttpClientErrorException.Unauthorized)
                exceptionMessage = "Unauthorized";
            else if (e instanceof HttpServerErrorException.InternalServerError)
                exceptionMessage = "InternalServerError";
            else if (e instanceof HttpServerErrorException.NotImplemented)
                exceptionMessage = "NotImplemented";
            else if (e instanceof HttpServerErrorException.ServiceUnavailable)
                exceptionMessage = "ServiceUnavailable";
            else exceptionMessage = String.format("Unknown Error:%s", e.getMessage());

            return new RestWebInvocationResult(false, exceptionMessage, invocation, e.getRawStatusCode(),
                    new HashMap<>(), null);
        } catch (RuntimeException rex) {
            return new RestWebInvocationResult(false, String.format("Unkown Error:%s", rex.getMessage()), invocation, -1, new HashMap<>(), null);
        }
    }

    private String autocompleteURL(String originURL) {
        if (originURL.startsWith("http") || originURL.startsWith("https")) return originURL;
        if (originURL.startsWith(contextPath))
            return String.format("http://127.0.0.1:%d%s", port, originURL);
        else if (originURL.startsWith("/"))
            return String.format("http://127.0.0.1:%d%s%s", port, contextPath, originURL);
        else
            return String.format("http://127.0.0.1:%d%s/%s", port, contextPath, originURL);
    }
}

File: restdoc-rpc-client/restdoc-restweb-client/src/main/kotlin/restdoc/client/restweb/RestWebInvokerImpl.kt
package restdoc.client.restweb

import org.springframework.core.env.Environment
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.http.ResponseEntity
import org.springframework.web.client.HttpClientErrorException
import org.springframework.web.client.HttpServerErrorException
import org.springframework.web.client.RestClientException
import org.springframework.web.client.RestTemplate
import restdoc.client.api.Invoker
import restdoc.client.api.model.InvocationResult
import restdoc.client.api.model.RestWebInvocation
import restdoc.client.api.model.RestWebInvocationResult

/**
 * RestWebInvokerImpl
 */
class RestWebInvokerImpl(environment: Environment, private val restTemplate: RestTemplate) : Invoker<RestWebInvocation> {

    private val port: Int = environment.getProperty("server.port", "8080").toInt()
    private val contextPath: String = environment.getProperty("server.servlet.context-path", "")

    override fun rpcInvoke(t: RestWebInvocation): InvocationResult {
        val url = autocompleteURL(t.url)
        val requestHeaders = HttpHeaders()

        t.requestHeaders.forEach { (k, v) -> requestHeaders.addAll(k, v) }
        val httpEntity = HttpEntity(t.requestBody, requestHeaders)

        val responseEntity: ResponseEntity<Any>?
        return try {
            responseEntity = restTemplate.exchange(url, HttpMethod.valueOf(t.method), httpEntity, Any::class.java, t.uriVariable)

            RestWebInvocationResult(true, null,
                    t, responseEntity.statusCodeValue,
                    mutableMapOf(),
                    responseEntity.body)

        } catch (e: RestClientException) {
            when (e) {
                is HttpServerErrorException.BadGateway -> RestWebInvocationResult(false, "BadGateway", t, e.rawStatusCode, mutableMapOf(), null)
                is HttpClientErrorException.BadRequest -> RestWebInvocationResult(false, "BadRequest", t, e.rawStatusCode, mutableMapOf(), null)
                is HttpClientErrorException.Conflict -> RestWebInvocationResult(false, "Conflict", t, e.rawStatusCode, mutableMapOf(), null)
                is HttpClientErrorException.Forbidden -> RestWebInvocationResult(false, "Forbidden", t, e.rawStatusCode, mutableMapOf(), null)
                is HttpServerErrorException.GatewayTimeout -> RestWebInvocationResult(false, "GatewayTimeout", t, e.rawStatusCode, mutableMapOf(), null)
                is HttpClientErrorException.Gone -> RestWebInvocationResult(false, "Gone", t, e.rawStatusCode, mutableMapOf(), null)
                is HttpClientErrorException.NotFound -> RestWebInvocationResult(false, "NotFound", t, e.rawStatusCode, mutableMapOf(), null)
                is HttpClientErrorException.MethodNotAllowed -> RestWebInvocationResult(false, "MethodNotAllowed", t, e.rawStatusCode, mutableMapOf(), null)
                is HttpClientErrorException.NotAcceptable -> RestWebInvocationResult(false, "NotAcceptable", t, e.rawStatusCode, mutableMapOf(), null)
                is HttpClientErrorException.UnsupportedMediaType -> RestWebInvocationResult(false, "UnsupportedMediaType", t, e.rawStatusCode, mutableMapOf(), null)
                is HttpClientErrorException.UnprocessableEntity -> RestWebInvocationResult(false, "UnprocessableEntity", t, e.rawStatusCode, mutableMapOf(), null)
                is HttpClientErrorException.TooManyRequests -> RestWebInvocationResult(false, "TooManyRequests", t, e.rawStatusCode, mutableMapOf(), null)
                is HttpClientErrorException.Unauthorized -> RestWebInvocationResult(false, "Unauthorized", t, e.rawStatusCode, mutableMapOf(), null)
                is HttpServerErrorException.InternalServerError -> RestWebInvocationResult(false, "InternalServerError", t, e.rawStatusCode, mutableMapOf(), null)
                is HttpServerErrorException.NotImplemented -> RestWebInvocationResult(false, "NotImplemented", t, e.rawStatusCode, mutableMapOf(), null)
                is HttpServerErrorException.ServiceUnavailable -> RestWebInvocationResult(false, "ServiceUnavailable", t, e.rawStatusCode, mutableMapOf(), null)
                else -> RestWebInvocationResult(false, "未知错误${e.message}", t, -1, mutableMapOf(), null)
            }
        } catch (e: RuntimeException) {
            RestWebInvocationResult(false, "未知错误${e.message}", t, -1, mutableMapOf(), null)
        }
    }

    private fun autocompleteURL(originURL: String): String {
        return if (originURL.startsWith("http") || originURL.startsWith("https")) originURL
        else {
            if (originURL.startsWith(contextPath)) {
                String.format("http://127.0.0.1:%d%s", port, originURL)
            } else {
                if (originURL.startsWith("/")) {
                    String.format("http://127.0.0.1:%d%s%s", port, contextPath, originURL)
                } else {
                    String.format("http://127.0.0.1:%d%s/%s", port, contextPath, originURL)
                }
            }
        }
    }
}